src/stream/src/executor/over_window/state/buffer.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::VecDeque;
+
+use risingwave_expr::function::window::Frame;
+
+/// Actually with `VecDeque` as internal buffer, we don't need split key and value here. Just in
+/// case we want to switch to `BTreeMap` later, so that the general version of `OverWindow` executor
+/// can reuse this.
+struct Entry<K: Ord, V> {
+    key: K,
+    value: V,
+}
+
+pub(super) struct WindowBuffer<K: Ord, V> {
+    frame: Frame,
+    buffer: VecDeque<Entry<K, V>>,
+    curr_idx: usize,
+    right_idx: usize, // inclusive, note this can be less than `curr_idx`
+}
+
+const LEFT_IDX: usize = 0;
+
+/// Two simple properties of a window frame:
+/// 1. If the following half of the window is saturated, the current window is ready.
+/// 2. If the preceding half of the window is saturated, the first several entries can be removed
+/// when sliding.
+///
+/// Note: A window frame can be pure preceding, pure following, or across the _current row_.
+pub(super) struct CurrWindow<'a, K> {
+    pub key: Option<&'a K>,
+    pub preceding_saturated: bool,
+    pub following_saturated: bool,
+}
+
+impl<K> CurrWindow<'_, K> {
+    pub fn is_ready(&self) -> bool {
+        self.following_saturated
+    }
+}
+
+impl<K: Ord, V> WindowBuffer<K, V> {
+    pub fn new(frame: Frame) -> Self {
+        let right_idx = match &frame {
+            Frame::Rows(start, end) => {
+                let start_off = start.to_offset();
+                let end_off = end
+                    .to_offset()
+                    .expect("frame end cannot be unbounded in streaming mode");
+                if let Some(start_off) = start_off {
+                    assert!(start_off <= end_off);
+                }
+                if end_off <= 0 {
+                    0
```

Comment:
should it be `LEFT_IDX` here?

Reply:
Seems to have bug here, let me add some unit tests later.

Reply:
Fixed a bug related to index maintaining, PTAL